lib/pure/decimal.nim

```diff
@@ -0,0 +1,1040 @@
+  var part = hex[0 .. 7]
+  result.a = fromHex[uint32](part)
+  part = hex[8 .. 15]
+  result.b = fromHex[uint32](part)
+  part = hex[16 .. 23]
+  result.c = fromHex[uint32](part)
+  part = hex[24 .. 31]
+  result.d = fromHex[uint32](part)
+
+#
+# output
+#
+
+proc simpleDigitStr(dList: SignificandArray): string =
+  var firstDigitSeen = false
+  for digit in dList:
+    if digit != 0.byte:
+      firstDigitSeen = true
+    if firstDigitSeen:
+      result &= $(digit.int)
+  if not firstDigitSeen:
+    result = "0"
+
+proc generateNumberString(number: Decimal, exponent: int16, upper: uint32): string =
+  # from the lower three uint32 in decimal and the masked upper uint32 (upper),
+  # derive a sequence of decimal bytes.
+  if number.negative:
+    result = "-"
+  else:
+    result = ""
+
+  let digits = generateDigits(number, upper)
+  let justDigits = simpleDigitStr(digits)
+
+  let scientificExponent = justDigits.len - 1 + exponent
+  if (scientificExponent < -6) or (exponent > 0):
+    # express with scientific notation
+    for index, ch in justDigits:
+      if index == 1:
+        result &= "."
+      result &= ch
+    result &= "E"
+    if scientificExponent >= 0:
+      result &= "+"
+    result &= $scientificExponent
+  elif exponent == 0:
+    # if zero decimal places, then it is a simple integer
+    result &= justDigits
+  else:
+    let significance = -exponent
+    let leadingZeroes = significance - justDigits.len
+    if leadingZeroes >= 0:
+      result &= "0."
+      result &= "0".repeat(leadingZeroes)
+      result &= justDigits
+    else:
+      let depth = -leadingZeroes
+      for index, ch in justDigits:
+        if index == depth:
+          result &= "."
+        result &= ch
+
+proc sci*(number: Decimal): string =
+  ## Express the Decimal value in Scientific Notation
+  var (decKind, exponent, upper) = determineKindExponentAndUpper(number)
+  case deckind:
+  of dkValued:
+    let digits = generateDigits(number, upper)
+    let justDigits = simpleDigitStr(digits)
+    let scientificExponent = justDigits.len - 1 + exponent
+    for index, ch in justDigits:
+      if index == 1:
+        result &= "."
+      result &= ch
+    result &= "E"
+    if scientificExponent >= 0:
+      result &= "+"
+    result &= $scientificExponent
+  of dkInfinite:
+    if number.negative:
+      result = "-Infinity"
+    else:
+      result = "Infinity"
+  of dkNaN:
+    result = "NaN"
+
+proc `$`*(number: Decimal): string =
+  ## Express the Decimal value as a canonical string
+  var (decKind, exponent, upper) = determineKindExponentAndUpper(number)
+  case deckind:
+  of dkValued:
+    result &= generateNumberString(number, exponent, upper)
+  of dkInfinite:
+    if number.negative:
+      result = "-Infinity"
+    else:
+      result = "Infinity"
+  of dkNaN:
+    result = "NaN"
+
+proc internalRepr*(number: Decimal): string =
+  result = "Decimal($1 $2 $3 $4)".format(number.a, number.b, number.c, number.d)
```

Comment:
```suggestion
proc internalRepr*(number: Decimal): string =
  "Decimal($1 $2 $3 $4)".format(number.a, number.b, number.c, number.d)
```
Nit: you can leave out the `result = ` when it's the last statement (ditto for other cases). This is subjective though (there's nothing about it in NEP1 afaik).

Reply:
fixed in next commit.